internal/pkg/aws/codepipeline/codepipeline.go

```diff
@@ -210,6 +212,39 @@ func (c *CodePipeline) ListPipelineNamesByTags(tags map[string]string) ([]string
 	return pipelineNames, nil
 }
 
+// ListPipelineExecution returns the ExecutionID of the most recent execution of a pipeline.
+func (c *CodePipeline) ListPipelineExecution(pipelineName string) (*string, error) {
+	input := &cp.ListPipelineExecutionsInput{
+		MaxResults:   aws.Int64(1),
+		PipelineName: &pipelineName,
+	}
+	output, err := c.client.ListPipelineExecutions(input)
+	if err != nil {
+		return nil, fmt.Errorf("list pipeline execution for %s: %w", pipelineName, err)
+	}
+	return output.PipelineExecutionSummaries[0].PipelineExecutionId, nil
+}
+
+// RetryStageExecution tries to rebuild a failed pipeline.
+func (c *CodePipeline) RetryStageExecution(pipelineName string) error {
+	executionID, err := c.ListPipelineExecution(pipelineName)
+	if err != nil {
+		return fmt.Errorf("retrieve pipeline execution ID: %w", err)
+	}
+	input :=
+		&cp.RetryStageExecutionInput{
+			PipelineExecutionId: executionID,
+			PipelineName:        &pipelineName,
+			RetryMode:           aws.String("FAILED_ACTIONS"),
+			StageName:           aws.String("Source"),
```

Comment:
Can this be a const so we can change it more easily if the source stage name is ever modified?

Reply:
The aim of `internal/pkg/aws` pkgs is to provide utility functions on top of the AWS SDK, so we should avoid any sort of Copilot specific entities as inputs or within the body.

The stage name "Source" is something unique to Copilot pipelines, so my suggestion would be to instead have this method accept a second parameter `RetryStageExecution(pipelineName, stageName string)` to make it flexible to retry any stage.

And then in the `deploy/cloudformation` pkg we can have a const like Austin suggested above for the source stage that's passed as input.

Reply:
Done! Thanks, all.